qutebrowser/browser/webengine/notification.py

```diff
@@ -0,0 +1,138 @@
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Handles sending notifications over DBus."""
+
+import typing
+
+from qutebrowser.utils import log
+
+from PyQt5.QtGui import QImage
+from PyQt5.QtCore import QVariant, QMetaType, QByteArray, PYQT_VERSION
+from PyQt5.QtDBus import QDBusConnection, QDBusInterface, QDBus, QDBusArgument
+from PyQt5.QtWebEngineCore import QWebEngineNotification
+from PyQt5.QtWebEngineWidgets import QWebEngineProfile
+
+
+class DBusException(Exception):
+    """Raised when something goes wrong with talking to DBus."""
+
+
+class DBusNotificationPresenter:
+    """Manages notifications that are sent over DBus."""
+
+    def __init__(self):
+        bus = QDBusConnection.sessionBus()
+        if not bus.isConnected():
+            raise DBusException("Failed to connect to DBus session bus")
+
+        self.interface = QDBusInterface(
+            "org.freedesktop.Notifications",  # service
+            "/org/freedesktop/Notifications",  # path
+            "org.freedesktop.Notifications",  # interface
+            bus,
+        )
+
+        if not self.interface:
+            raise DBusException("Could not construct a DBus interface")
+
+    def install(self, profile: QWebEngineProfile) -> None:
+        """Sets the profile to use the manager as the presenter."""
+        # WORKAROUND for https://www.riverbankcomputing.com/pipermail/pyqt/2020-May/042916.html
+        if PYQT_VERSION < 0x050F00:  # PyQt 5.15
+            # PyQtWebEngine unrefs the callback after it's called, for some reason.
+            # So we call setNotificationPresenter again to *increase* its refcount
+            # to prevent it from getting GC'd. Otherwise, random methods start
+            # getting called with the notification as `self`, or segfaults happen,
+            # or other badness.
+            def _present_and_reset(qt_notification: QWebEngineNotification) -> None:
+                profile.setNotificationPresenter(_present_and_reset)
+                self._present(qt_notification)
+            profile.setNotificationPresenter(_present_and_reset)
+        else:
+            profile.setNotificationPresenter(self._present)
+
+    def _present(self, qt_notification: QWebEngineNotification) -> None:
+        """Shows a notification over DBus.
+
+        This should *not* be directly passed to setNotificationPresenter
+        because of a bug in the PyQtWebEngine bindings.
```

Comment:
Maybe update this to say "On Qt < 5.15" or so?

Reply:
Done. There's some subtlety here since it works on Qt 5.14, but not on *PyQtWebEngine* 5.14. Maybe there needs to be a `qtutils.version_check` equivalent for that?

Reply:
There certainly should be! See #3785.